File: src/symphonic_cipher/scbe_aethermoore/ai_brain/tri_manifold_lattice.py
"""
Tri-Manifold Lattice — Temporal Harmonic Governance (Python Reference)
=====================================================================

Three temporal manifolds over the Poincaré ball with harmonic scaling
H(d, R) = R^(d²) for super-exponential cost amplification.

Components:
  - TemporalWindow: Sliding window for hyperbolic distance averaging
  - harmonicScale: H(d, R) = R^(d²) — super-exponential amplification
  - triadicDistance: Weighted Euclidean norm of 3 manifold distances
  - TriManifoldLattice: Full lattice manager with resonance/anomaly detection

Mathematical Properties:
  - d_tri ≥ 0 (non-negativity)
  - d_tri = 0 ⟺ d₁ = d₂ = d_G = 0 (positive-definiteness)
  - ∂d_tri/∂dᵢ = λᵢ·dᵢ / d_tri ≥ 0 (monotonicity)
  - H(d, R) × H(d, 1/R) = 1 (duality / phase cancellation)
  - H grows super-exponentially: faster than any single exponential

@module ai_brain/tri_manifold_lattice
@layer Layer 5, Layer 11, Layer 12, Layer 14
@version 1.0.0
"""

import math
import time
from collections import deque
from dataclasses import dataclass, field
from typing import Dict, List, Optional, Tuple

# Import from sibling modules
from .unified_state import (
    BRAIN_DIMENSIONS,
    BRAIN_EPSILON,
    PHI,
    hyperbolic_distance_safe,
    safe_poincare_embed,
    _vec_norm as vector_norm,
)

# ═══════════════════════════════════════════════════════════════
# Constants
# ═══════════════════════════════════════════════════════════════

HARMONIC_R = 1.5
"""Default harmonic ratio: perfect fifth (3:2) from Pythagorean tuning."""

DEFAULT_WINDOW_IMMEDIATE = 5
DEFAULT_WINDOW_MEMORY = 25
DEFAULT_WINDOW_GOVERNANCE = 100

MAX_LATTICE_DEPTH = 1000


@dataclass
class TriadicWeights:
    """Weights for the triadic distance (must sum to 1)."""

    immediate: float = 0.5
    memory: float = 0.3
    governance: float = 0.2

    def normalized(self) -> "TriadicWeights":
        """Return normalized copy that sums to 1."""
        s = self.immediate + self.memory + self.governance
        if s <= 0:
            return TriadicWeights(1 / 3, 1 / 3, 1 / 3)
        return TriadicWeights(self.immediate / s, self.memory / s, self.governance / s)


@dataclass
class LatticeNode:
    """A single lattice node capturing the full tri-manifold state."""

    tick: int
    raw_state: List[float]
    embedded: List[float]
    hyperbolic_dist: float
    manifold_distances: Dict[str, float]
    triadic_distance: float
    harmonic_cost: float
    embedded_norm: float
    timestamp: float


@dataclass
class LatticeSnapshot:
    """Tri-manifold lattice snapshot for external consumption."""

    tick: int
    triadic_distance: float
    harmonic_cost: float
    manifold_distances: Dict[str, float]
    weights: TriadicWeights
    node_count: int
    drift_velocity: float


# ═══════════════════════════════════════════════════════════════
# Harmonic Scaling Law
# ═══════════════════════════════════════════════════════════════


def harmonic_scale(d: int, R: float = HARMONIC_R) -> float:
    """Harmonic Scaling: H(d, R) = R^(d²).

    Super-exponential amplification where each dimension multiplies
    complexity via pairwise interactions (d² exponent).

    Args:
        d: Number of dimensions (typically 1-6).
        R: Harmonic ratio (default: 1.5, the perfect fifth).

    Returns:
        R^(d²) — the amplification factor.
    """
    if d < 0:
        return 1.0
    if R <= 0:
        return 0.0
    return R ** (d * d)


def harmonic_scale_inverse(d: int, R: float = HARMONIC_R) -> float:
    """Inverse harmonic scaling: H(d, 1/R) = (1/R)^(d²).

    Property: harmonic_scale(d, R) * harmonic_scale_inverse(d, R) = 1
    """
    if d < 0 or R <= 0:
        return 1.0
    return (1.0 / R) ** (d * d)


def harmonic_scale_table(
    max_d: int, R: float = HARMONIC_R
) -> List[Dict[str, float]]:
    """Harmonic scaling table for dimensions 1..max_d."""
    table = []
    for d in range(1, max_d + 1):
        s = harmonic_scale(d, R)
        table.append({"d": d, "scale": s, "log_scale": math.log(s) if s > 0 else 0})
    return table


# ═══════════════════════════════════════════════════════════════
# Triadic Distance
# ═══════════════════════════════════════════════════════════════


def triadic_distance(
    d1: float,
    d2: float,
    d_g: float,
    weights: Optional[TriadicWeights] = None,
) -> float:
    """Triadic temporal distance: weighted Euclidean norm of 3 manifold distances.

    d_tri = √(λ₁·d₁² + λ₂·d₂² + λ₃·d_G²)

    Args:
        d1: Immediate manifold distance.
        d2: Memory manifold distance.
        d_g: Governance manifold distance.
        weights: Triadic weights (must sum to 1).

    Returns:
        Combined triadic distance (non-negative).
    """
    if weights is None:
        weights = TriadicWeights()
    sum_sq = weights.immediate * d1 * d1 + weights.memory * d2 * d2 + weights.governance * d_g * d_g
    return math.sqrt(max(0.0, sum_sq))


def triadic_partial(d_i: float, lambda_i: float, d_tri: float) -> float:
    """Partial derivative of triadic distance w.r.t. component i.

    ∂d_tri/∂dᵢ = λᵢ·dᵢ / d_tri
    """
    if d_tri < BRAIN_EPSILON:
        return 0.0
    return (lambda_i * d_i) / d_tri


# ═══════════════════════════════════════════════════════════════
# Temporal Window
# ═══════════════════════════════════════════════════════════════


class TemporalWindow:
    """Sliding window for temporal manifold distance averaging.

    Maintains a fixed-size circular buffer of hyperbolic distances
    and provides the windowed average d_k(t) = (1/W_k) Σ d_H(u(s), ℓ).
    """

    def __init__(self, size: int):
        if size < 1:
            raise ValueError("Window size must be >= 1")
        self.size = size
        self._buffer: deque = deque(maxlen=size)
        self._sum: float = 0.0

    def push(self, distance: float) -> None:
        """Push a new distance sample into the window."""
        if len(self._buffer) >= self.size:
            self._sum -= self._buffer[0]
        self._buffer.append(distance)
        self._sum += distance

    def average(self) -> float:
        """Windowed average distance."""
        if len(self._buffer) == 0:
            return 0.0
        return self._sum / len(self._buffer)

    def filled(self) -> int:
        """Current sample count."""
        return len(self._buffer)

    def is_warmed_up(self) -> bool:
        """Whether the window is fully warmed up."""
        return len(self._buffer) >= self.size

    def latest(self) -> float:
        """Most recent distance sample."""
        if len(self._buffer) == 0:
            return 0.0
        return self._buffer[-1]

    def variance(self) -> float:
        """Variance of distances in the window."""
        n = len(self._buffer)
        if n < 2:
            return 0.0
        avg = self.average()
        return sum((x - avg) ** 2 for x in self._buffer) / (n - 1)

    def reset(self) -> None:
        """Reset window to empty state."""
        self._buffer.clear()
        self._sum = 0.0


# ═══════════════════════════════════════════════════════════════
# Tri-Manifold Lattice
# ═══════════════════════════════════════════════════════════════


class TriManifoldLattice:
    """Tri-Manifold Lattice: three temporal manifolds over the Poincaré ball.

    Each manifold samples hyperbolic distance at a different timescale,
    then the triadic distance combines them into a single governance metric.
    The harmonic scaling law amplifies this across dimensional space.

    Usage:
        lattice = TriManifoldLattice()
        node = lattice.ingest(state_vector)
        print(node.triadic_distance)   # Combined temporal distance
        print(node.harmonic_cost)      # Super-exponential governance cost
    """

    def __init__(
        self,
        *,
        window_immediate: int = DEFAULT_WINDOW_IMMEDIATE,
        window_memory: int = DEFAULT_WINDOW_MEMORY,
        window_governance: int = DEFAULT_WINDOW_GOVERNANCE,
        weights: Optional[TriadicWeights] = None,
        harmonic_r: float = HARMONIC_R,
        harmonic_dimensions: int = 6,
        reference_point: Optional[List[float]] = None,
    ):
        self._immediate = TemporalWindow(window_immediate)
        self._memory = TemporalWindow(window_memory)
        self._governance = TemporalWindow(window_governance)

        self._weights = (weights or TriadicWeights()).normalized()
        self._harmonic_r = harmonic_r
        self._harmonic_dimensions = harmonic_dimensions
        self._reference = reference_point or [0.0] * BRAIN_DIMENSIONS

        self._nodes: List[LatticeNode] = []
        self._tick: int = 0

    def ingest(self, raw_state: List[float]) -> LatticeNode:
        """Ingest a new 21D state vector into the lattice.

        1. Embeds into Poincaré ball
        2. Computes hyperbolic distance from reference
        3. Pushes to all three temporal windows
        4. Computes triadic distance and harmonic cost

        Returns:
            LatticeNode with all computed metrics.
        """
        self._tick += 1

        embedded = safe_poincare_embed(raw_state)
        ref_embedded = safe_poincare_embed(self._reference)
        embedded_norm = vector_norm(embedded)

        h_dist = hyperbolic_distance_safe(embedded, ref_embedded)

        self._immediate.push(h_dist)
        self._memory.push(h_dist)
        self._governance.push(h_dist)

        d1 = self._immediate.average()
        d2 = self._memory.average()
        d_g = self._governance.average()

        d_tri = triadic_distance(d1, d2, d_g, self._weights)

        h_scale = harmonic_scale(self._harmonic_dimensions, self._harmonic_r)
        h_cost = d_tri * h_scale

        node = LatticeNode(
            tick=self._tick,
            raw_state=list(raw_state),
            embedded=embedded,
            hyperbolic_dist=h_dist,
            manifold_distances={"immediate": d1, "memory": d2, "governance": d_g},
            triadic_distance=d_tri,
            harmonic_cost=h_cost,
            embedded_norm=embedded_norm,
            timestamp=time.time(),
        )

        self._nodes.append(node)
        if len(self._nodes) > MAX_LATTICE_DEPTH:
            self._nodes = self._nodes[-MAX_LATTICE_DEPTH:]

        return node

    def drift_velocity(self) -> float:
        """Rate of change of triadic distance (finite difference)."""
        if len(self._nodes) < 2:
            return 0.0
        return self._nodes[-1].triadic_distance - self._nodes[-2].triadic_distance

    def drift_acceleration(self) -> float:
        """Second derivative of triadic distance."""
        if len(self._nodes) < 3:
            return 0.0
        d2 = self._nodes[-1].triadic_distance
        d1 = self._nodes[-2].triadic_distance
        d0 = self._nodes[-3].triadic_distance
        return d2 - 2 * d1 + d0

    def current_triadic_distance(self) -> float:
        """Current triadic distance (0 if no samples)."""
        if not self._nodes:
            return 0.0
        return self._nodes[-1].triadic_distance

    def current_harmonic_cost(self) -> float:
        """Current harmonic cost (0 if no samples)."""
        if not self._nodes:
            return 0.0
        return self._nodes[-1].harmonic_cost

    def temporal_resonance(self) -> float:
        """Temporal resonance coefficient in [0, 1].

        1 = perfect agreement across all three manifolds.
        Lower values indicate temporal drift/divergence.
        """
        if not self._nodes:
            return 1.0
        latest = self._nodes[-1]
        d1 = latest.manifold_distances["immediate"]
        d2 = latest.manifold_distances["memory"]
        d_g = latest.manifold_distances["governance"]
        avg = (d1 + d2 + d_g) / 3.0
        if avg < BRAIN_EPSILON:
            return 1.0
        variance = ((d1 - avg) ** 2 + (d2 - avg) ** 2 + (d_g - avg) ** 2) / 3.0
        return 1.0 / (1.0 + variance / (avg * avg))

    def temporal_anomaly(self) -> float:
        """Temporal anomaly: |d₁ - d_G| / d_G.

        High values indicate immediate window diverges from governance.
        """
        if not self._nodes:
            return 0.0
        latest = self._nodes[-1]
        d1 = latest.manifold_distances["immediate"]
        d_g = latest.manifold_distances["governance"]
        denom = max(d_g, BRAIN_EPSILON)
        return abs(d1 - d_g) / denom

    def snapshot(self) -> LatticeSnapshot:
        """Current lattice snapshot."""
        latest = self._nodes[-1] if self._nodes else None
        return LatticeSnapshot(
            tick=self._tick,
            triadic_distance=latest.triadic_distance if latest else 0.0,
            harmonic_cost=latest.harmonic_cost if latest else 0.0,
            manifold_distances=latest.manifold_distances if latest else {"immediate": 0, "memory": 0, "governance": 0},
            weights=TriadicWeights(self._weights.immediate, self._weights.memory, self._weights.governance),
            node_count=len(self._nodes),
            drift_velocity=self.drift_velocity(),
        )

    def verify_duality(self, d: int) -> Tuple[float, float, float]:
        """Verify H(d, R) * H(d, 1/R) = 1."""
        fwd = harmonic_scale(d, self._harmonic_r)
        inv = harmonic_scale_inverse(d, self._harmonic_r)
        return fwd, inv, fwd * inv

    @property
    def tick(self) -> int:
        return self._tick

    @property
    def weights(self) -> TriadicWeights:
        return self._weights

    def reset(self) -> None:
        """Reset all windows and lattice state."""
        self._immediate.reset()
        self._memory.reset()
        self._governance.reset()
        self._nodes.clear()
        self._tick = 0

File: src/ai_brain/tri-manifold-lattice.ts
/**
 * @file tri-manifold-lattice.ts
 * @module ai_brain/tri-manifold-lattice
 * @layer Layer 5, Layer 11, Layer 12, Layer 14
 * @component Tri-Manifold Lattice — Temporal Harmonic Governance
 * @version 1.0.0
 * @since 2026-02-10
 *
 * Constructs a lattice over THREE temporal manifolds (immediate, memory,
 * governance), each sampling hyperbolic distance in the Poincaré ball at
 * different timescales. The triadic distance combines them via weighted
 * Euclidean norm, and the Harmonic Scaling Law H(d, R) = R^(d²) applies
 * super-exponential cost amplification across dimensions.
 *
 * Architecture:
 *
 *   Manifold 1 (Immediate):  Window W₁ — short-term drift detection
 *   Manifold 2 (Memory):     Window W₂ — medium-term pattern memory
 *   Manifold 3 (Governance): Window W_G — long-term policy enforcement
 *
 *   Triadic Distance:
 *     d_tri(t) = √(λ₁·d₁² + λ₂·d₂² + λ₃·d_G²)
 *     where λᵢ ≥ 0 and Σλᵢ = 1
 *
 *   Harmonic Scaling (super-exponential):
 *     H(d, R) = R^(d²)
 *     For R = 1.5 (perfect fifth):
 *       d=1: 1.5,  d=2: 5.06,  d=3: 38.44,  d=4: 656.84,  d=6: 2,184,164
 *
 *   Lattice Node: a point in the tri-manifold with:
 *     - 21D Poincaré embedding
 *     - Three temporal distance averages
 *     - Triadic distance (combined scalar)
 *     - Harmonic-scaled governance cost
 *
 * Properties:
 *   - d_tri is a weighted Euclidean norm: non-negative, positive-definite
 *   - d_tri = 0 ⟺ d₁ = d₂ = d_G = 0 (all windows report zero drift)
 *   - ∂d_tri/∂dᵢ = λᵢ·dᵢ / d_tri ≥ 0 (monotonic in each component)
 *   - H(d, R) × H(d, 1/R) = 1 (duality / phase cancellation)
 *   - H is super-exponential: grows faster than any single exponential
 *
 * Integration:
 *   - Uses hyperbolicDistanceSafe() from unified-state.ts
 *   - Uses safePoincareEmbed() for 21D → B²¹ embedding
 *   - Feeds into Hamiltonian Braid for governance decisions
 *   - Audio axis (L14) coherence modulates governance window
 */

import { PHI, BRAIN_EPSILON, BRAIN_DIMENSIONS } from './types.js';
import { hyperbolicDistanceSafe, safePoincareEmbed, vectorNorm } from './unified-state.js';

// ═══════════════════════════════════════════════════════════════
// Constants
// ═══════════════════════════════════════════════════════════════

/** Default harmonic ratio: perfect fifth (3:2) from Pythagorean tuning */
export const HARMONIC_R = 1.5;

/** Default temporal window sizes (in ticks/samples) */
export const DEFAULT_WINDOW_SIZES = {
  immediate: 5,    // W₁: last 5 samples — fast reaction
  memory: 25,      // W₂: last 25 samples — pattern memory
  governance: 100,  // W_G: last 100 samples — policy enforcement
} as const;

/** Default triadic weights (λ₁, λ₂, λ₃), sum to 1 */
export const DEFAULT_TRIADIC_WEIGHTS: TriadicWeights = {
  immediate: 0.5,    // λ₁: fast signals weighted highest
  memory: 0.3,       // λ₂: medium-term patterns
  governance: 0.2,   // λ₃: slow governance drift
};

/** Maximum lattice depth before pruning old nodes */
export const MAX_LATTICE_DEPTH = 1000;

// ═══════════════════════════════════════════════════════════════
// Types
// ═══════════════════════════════════════════════════════════════

/** Triadic weight configuration (must sum to 1) */
export interface TriadicWeights {
  immediate: number;
  memory: number;
  governance: number;
}

/** Configuration for the tri-manifold lattice */
export interface TriManifoldConfig {
  windowSizes?: {
    immediate?: number;
    memory?: number;
    governance?: number;
  };
  weights?: Partial<TriadicWeights>;
  harmonicR?: number;
  /** Reference point in Poincaré ball (default: origin) */
  referencePoint?: number[];
  /** Number of harmonic dimensions for scaling (default: 6, one per tongue) */
  harmonicDimensions?: number;
}

/** A single lattice node capturing the full tri-manifold state */
export interface LatticeNode {
  /** Monotonic tick counter */
  tick: number;
  /** Raw 21D state vector */
  rawState: number[];
  /** Poincaré ball embedding */
  embedded: number[];
  /** Hyperbolic distance from reference point */
  hyperbolicDist: number;
  /** Windowed average distances per manifold */
  manifoldDistances: {
    immediate: number;  // d₁(t)
    memory: number;     // d₂(t)
    governance: number; // d_G(t)
  };
  /** Combined triadic distance */
  triadicDistance: number;
  /** Harmonic-scaled governance cost */
  harmonicCost: number;
  /** Norm of the Poincaré embedding (boundary proximity) */
  embeddedNorm: number;
  /** Wall-clock timestamp */
  timestamp: number;
}

/** Tri-manifold lattice snapshot for external consumption */
export interface LatticeSnapshot {
  tick: number;
  triadicDistance: number;
  harmonicCost: number;
  manifoldDistances: {
    immediate: number;
    memory: number;
    governance: number;
  };
  weights: TriadicWeights;
  nodeCount: number;
  driftVelocity: number;
}

// ═══════════════════════════════════════════════════════════════
// Harmonic Scaling Law
// ═══════════════════════════════════════════════════════════════

/**
 * Harmonic Scaling: H(d, R) = R^(d²)
 *
 * Super-exponential amplification where each dimension multiplies
 * complexity via pairwise interactions (d² exponent).
 *
 * Physical roots: helioseismology (solar oscillations scale as l²),
 * cymatics (Chladni plate modes scale as (m+2n)²).
 *
 * @param d - Number of dimensions (typically 1-6)
 * @param R - Harmonic ratio (default: 1.5, the perfect fifth)
 * @returns R^(d²) — the amplification factor
 */
export function harmonicScale(d: number, R: number = HARMONIC_R): number {
  if (d < 0) return 1; // No negative dimensions
  if (R <= 0) return 0; // Degenerate ratio
  return Math.pow(R, d * d);
}

/**
 * Inverse harmonic scaling for flux management / phase cancellation.
 * H(d, 1/R) = (1/R)^(d²)
 *
 * Property: harmonicScale(d, R) * harmonicScaleInverse(d, R) = 1
 */
export function harmonicScaleInverse(d: number, R: number = HARMONIC_R): number {
  if (d < 0 || R <= 0) return 1;
  return Math.pow(1 / R, d * d);
}

/**
 * Harmonic scaling table for dimensions 1..maxD.
 * Useful for precomputation and visualization.
 */
export function harmonicScaleTable(
  maxD: number,
  R: number = HARMONIC_R,
): Array<{ d: number; scale: number; logScale: number }> {
  const table: Array<{ d: number; scale: number; logScale: number }> = [];
  for (let d = 1; d <= maxD; d++) {
    const s = harmonicScale(d, R);
    table.push({ d, scale: s, logScale: Math.log(s) });
  }
  return table;
}

// ═══════════════════════════════════════════════════════════════
// Triadic Distance
// ═══════════════════════════════════════════════════════════════

/**
 * Triadic temporal distance: weighted Euclidean norm of 3 manifold distances.
 *
 * d_tri = √(λ₁·d₁² + λ₂·d₂² + λ₃·d_G²)
 *
 * Properties:
 *   - Non-negative (sum of squares under sqrt)
 *   - Positive-definite: d_tri = 0 ⟺ all dᵢ = 0
 *   - Monotonic in each component: ∂d_tri/∂dᵢ ≥ 0
 *
 * @param d1 - Immediate manifold distance
 * @param d2 - Memory manifold distance
 * @param dG - Governance manifold distance
 * @param weights - Triadic weights (must sum to 1)
 */
export function triadicDistance(
  d1: number,
  d2: number,
  dG: number,
  weights: TriadicWeights = DEFAULT_TRIADIC_WEIGHTS,
): number {
  const sumSq =
    weights.immediate * d1 * d1 +
    weights.memory * d2 * d2 +
    weights.governance * dG * dG;
  return Math.sqrt(Math.max(0, sumSq));
}

/**
 * Partial derivative of triadic distance w.r.t. component i.
 * ∂d_tri/∂dᵢ = λᵢ·dᵢ / d_tri
 */
export function triadicPartial(
  dI: number,
  lambdaI: number,
  dTri: number,
): number {
  if (dTri < BRAIN_EPSILON) return 0;
  return (lambdaI * dI) / dTri;
}

// ═══════════════════════════════════════════════════════════════
// Temporal Window (Sliding Average)
// ═══════════════════════════════════════════════════════════════

/**
 * Sliding window for temporal manifold distance averaging.
 *
 * Maintains a fixed-size circular buffer of hyperbolic distances
 * and provides the windowed average d_k(t) = (1/W_k) Σ d_H(u(s), ℓ)
 */
export class TemporalWindow {
  private buffer: number[];
  private head: number = 0;
  private count: number = 0;
  private sum: number = 0;

  constructor(public readonly size: number) {
    if (size < 1) throw new Error('Window size must be >= 1');
    this.buffer = new Array(size).fill(0);
  }

  /** Push a new distance sample into the window. */
  push(distance: number): void {
    if (this.count >= this.size) {
      // Remove oldest value
      this.sum -= this.buffer[this.head];
    }
    this.buffer[this.head] = distance;
    this.sum += distance;
    this.head = (this.head + 1) % this.size;
    if (this.count < this.size) this.count++;
  }

  /** Windowed average distance. */
  average(): number {
    if (this.count === 0) return 0;
    return this.sum / this.count;
  }

  /** Current sample count (may be < size during warmup). */
  filled(): number {
    return this.count;
  }

  /** Whether the window is fully warmed up. */
  isWarmedUp(): boolean {
    return this.count >= this.size;
  }

  /** Most recent distance sample. */
  latest(): number {
    if (this.count === 0) return 0;
    const idx = (this.head - 1 + this.size) % this.size;
    return this.buffer[idx];
  }

  /** Variance of distances in the window. */
  variance(): number {
    if (this.count < 2) return 0;
    const avg = this.average();
    let sumSq = 0;
    for (let i = 0; i < this.count; i++) {
      const diff = this.buffer[i] - avg;
      sumSq += diff * diff;
    }
    return sumSq / (this.count - 1);
  }

  /** Reset window to empty state. */
  reset(): void {
    this.buffer.fill(0);
    this.head = 0;
    this.count = 0;
    this.sum = 0;
  }
}

// ═══════════════════════════════════════════════════════════════
// Tri-Manifold Lattice
// ═══════════════════════════════════════════════════════════════

/**
 * Tri-Manifold Lattice: three temporal manifolds over the Poincaré ball.
 *
 * Each manifold samples hyperbolic distance at a different timescale,
 * then the triadic distance combines them into a single governance metric.
 * The harmonic scaling law amplifies this across dimensional space.
 *
 * Usage:
 *   const lattice = new TriManifoldLattice();
 *   const node = lattice.ingest(stateVector);
 *   console.log(node.triadicDistance);  // Combined temporal distance
 *   console.log(node.harmonicCost);     // Super-exponential governance cost
 */
export class TriManifoldLattice {
  // Three temporal manifold windows
  private readonly immediateWindow: TemporalWindow;
  private readonly memoryWindow: TemporalWindow;
  private readonly governanceWindow: TemporalWindow;

  // Configuration
  private readonly weights: TriadicWeights;
  private readonly harmonicR: number;
  private readonly harmonicDimensions: number;
  private readonly referencePoint: number[];

  // Lattice state
  private nodes: LatticeNode[] = [];
  private tick: number = 0;

  constructor(config?: TriManifoldConfig) {
    const ws = config?.windowSizes ?? {};
    this.immediateWindow = new TemporalWindow(ws.immediate ?? DEFAULT_WINDOW_SIZES.immediate);
    this.memoryWindow = new TemporalWindow(ws.memory ?? DEFAULT_WINDOW_SIZES.memory);
    this.governanceWindow = new TemporalWindow(ws.governance ?? DEFAULT_WINDOW_SIZES.governance);

    // Merge weights, normalize to sum=1
    const rawW = {
      immediate: config?.weights?.immediate ?? DEFAULT_TRIADIC_WEIGHTS.immediate,
      memory: config?.weights?.memory ?? DEFAULT_TRIADIC_WEIGHTS.memory,
      governance: config?.weights?.governance ?? DEFAULT_TRIADIC_WEIGHTS.governance,
    };
    const wSum = rawW.immediate + rawW.memory + rawW.governance;
    this.weights = {
      immediate: rawW.immediate / wSum,
      memory: rawW.memory / wSum,
      governance: rawW.governance / wSum,
    };

    this.harmonicR = config?.harmonicR ?? HARMONIC_R;
    this.harmonicDimensions = config?.harmonicDimensions ?? 6; // 6 tongues
    this.referencePoint = config?.referencePoint ?? new Array(BRAIN_DIMENSIONS).fill(0);
  }

  /**
   * Ingest a new 21D state vector into the lattice.
   *
   * 1. Embeds into Poincaré ball
   * 2. Computes hyperbolic distance from reference
   * 3. Pushes to all three temporal windows
   * 4. Computes triadic distance and harmonic cost
   *
   * @param rawState - 21D state vector (pre-Poincaré)
   * @returns The new lattice node with all computed metrics
   */
  ingest(rawState: number[]): LatticeNode {
    this.tick++;

    // Embed into Poincaré ball
    const embedded = safePoincareEmbed(rawState);
    const refEmbedded = safePoincareEmbed(this.referencePoint);
    const embeddedNorm = vectorNorm(embedded);

    // Hyperbolic distance from reference point
    const hDist = hyperbolicDistanceSafe(embedded, refEmbedded);

    // Push to all three temporal windows
    this.immediateWindow.push(hDist);
    this.memoryWindow.push(hDist);
    this.governanceWindow.push(hDist);

    // Windowed averages
    const d1 = this.immediateWindow.average();
    const d2 = this.memoryWindow.average();
    const dG = this.governanceWindow.average();

    // Triadic distance
    const dTri = triadicDistance(d1, d2, dG, this.weights);

    // Harmonic scaling: apply super-exponential cost
    // Scale triadic distance by H(dim, R) where dim = harmonicDimensions
    const hScale = harmonicScale(this.harmonicDimensions, this.harmonicR);
    const harmonicCost = dTri * hScale;

    const node: LatticeNode = {
      tick: this.tick,
      rawState: [...rawState],
      embedded,
      hyperbolicDist: hDist,
      manifoldDistances: {
        immediate: d1,
        memory: d2,
        governance: dG,
      },
      triadicDistance: dTri,
      harmonicCost,
      embeddedNorm,
      timestamp: Date.now(),
    };

    // Store node (with depth pruning)
    this.nodes.push(node);
    if (this.nodes.length > MAX_LATTICE_DEPTH) {
      this.nodes = this.nodes.slice(-MAX_LATTICE_DEPTH);
    }

    return node;
  }

  /**
   * Compute drift velocity: rate of change of triadic distance.
   * Uses finite difference of last two nodes.
   */
  driftVelocity(): number {
    if (this.nodes.length < 2) return 0;
    const curr = this.nodes[this.nodes.length - 1];
    const prev = this.nodes[this.nodes.length - 2];
    return curr.triadicDistance - prev.triadicDistance;
  }

  /**
   * Compute drift acceleration (second derivative).
   * Positive = drift is accelerating (concerning).
   */
  driftAcceleration(): number {
    if (this.nodes.length < 3) return 0;
    const n = this.nodes.length;
    const d2 = this.nodes[n - 1].triadicDistance;
    const d1 = this.nodes[n - 2].triadicDistance;
    const d0 = this.nodes[n - 3].triadicDistance;
    return (d2 - 2 * d1 + d0); // discrete second derivative
  }

  /** Get the current triadic distance (0 if no samples). */
  currentTriadicDistance(): number {
    if (this.nodes.length === 0) return 0;
    return this.nodes[this.nodes.length - 1].triadicDistance;
  }

  /** Get the current harmonic cost (0 if no samples). */
  currentHarmonicCost(): number {
    if (this.nodes.length === 0) return 0;
    return this.nodes[this.nodes.length - 1].harmonicCost;
  }

  /** Get the last N lattice nodes. */
  recentNodes(n: number): LatticeNode[] {
    return this.nodes.slice(-n);
  }

  /** Current lattice snapshot. */
  snapshot(): LatticeSnapshot {
    const latest = this.nodes[this.nodes.length - 1];
    return {
      tick: this.tick,
      triadicDistance: latest?.triadicDistance ?? 0,
      harmonicCost: latest?.harmonicCost ?? 0,
      manifoldDistances: latest?.manifoldDistances ?? { immediate: 0, memory: 0, governance: 0 },
      weights: { ...this.weights },
      nodeCount: this.nodes.length,
      driftVelocity: this.driftVelocity(),
    };
  }

  /**
   * Check if lattice is in "resonance" — all three manifolds agree.
   *
   * When immediate, memory, and governance distances are similar,
   * the system is in a stable temporal resonance (good or bad).
   * Divergence between manifolds indicates temporal drift.
   *
   * @returns Resonance coefficient in [0, 1] where 1 = perfect agreement
   */
  temporalResonance(): number {
    if (this.nodes.length === 0) return 1; // No data = trivially resonant
    const latest = this.nodes[this.nodes.length - 1];
    const { immediate: d1, memory: d2, governance: dG } = latest.manifoldDistances;
    const avg = (d1 + d2 + dG) / 3;
    if (avg < BRAIN_EPSILON) return 1; // All near zero = perfect resonance
    const variance =
      ((d1 - avg) ** 2 + (d2 - avg) ** 2 + (dG - avg) ** 2) / 3;
    // Map variance to [0, 1] resonance. Variance 0 → resonance 1.
    return 1 / (1 + variance / (avg * avg));
  }

  /**
   * Detect temporal anomaly: immediate window diverges from governance.
   *
   * If d₁ >> d_G, something just changed that hasn't been seen before.
   * If d_G >> d₁, a past pattern is repeating that's currently quiet.
   *
   * @returns Anomaly score in [0, ∞). Values > 2 are significant.
   */
  temporalAnomaly(): number {
    if (this.nodes.length === 0) return 0;
    const latest = this.nodes[this.nodes.length - 1];
    const { immediate: d1, governance: dG } = latest.manifoldDistances;
    const denom = Math.max(dG, BRAIN_EPSILON);
    return Math.abs(d1 - dG) / denom;
  }

  /**
   * Compute the harmonic scaling table for visualization.
   * Returns H(d, R) for d = 1..maxD.
   */
  harmonicTable(maxD: number = 6): Array<{ d: number; scale: number }> {
    return harmonicScaleTable(maxD, this.harmonicR);
  }

  /**
   * Check the harmonic duality property:
   * H(d, R) * H(d, 1/R) should equal 1 for all d.
   */
  verifyDuality(d: number): { forward: number; inverse: number; product: number } {
    const fwd = harmonicScale(d, this.harmonicR);
    const inv = harmonicScaleInverse(d, this.harmonicR);
    return { forward: fwd, inverse: inv, product: fwd * inv };
  }

  /** Total ticks ingested. */
  getTick(): number {
    return this.tick;
  }

  /** Current weights (normalized). */
  getWeights(): TriadicWeights {
    return { ...this.weights };
  }

  /** Reset all windows and lattice state. */
  reset(): void {
    this.immediateWindow.reset();
    this.memoryWindow.reset();
    this.governanceWindow.reset();
    this.nodes = [];
    this.tick = 0;
  }
}
